src/data/polynomial/high_order_derivative.lean

```diff
@@ -0,0 +1,153 @@
+import data.polynomial.derivative
+import algebra.ring.basic
+import data.polynomial.basic
+import logic.function.iterate
+import tactic.ring tactic.linarith
+
+noncomputable theory
+
+open finset nat polynomial
+open_locale big_operators
+
+namespace polynomial
+universes u
+variable {R : Type u}
+
+section semiring
+
+variables [semiring R] (f p q: polynomial R) (n k : ℕ)
+
+/--Defining the n-th derivative of polynomial f-/
+def deriv_n : polynomial R := derivative ^[n] f
+
+@[simp] lemma zeroth_deriv : deriv_n f 0 = f := rfl
+
+
+lemma deriv_n_succ : (deriv_n f n).derivative = (deriv_n f (n+1)) :=
+by rw [deriv_n, deriv_n, function.iterate_succ']
+
+lemma deriv_n_zero : deriv_n (0 : polynomial R) n = 0 :=
+begin
+  induction n with n hn;
+  simp only [deriv_n, id.def, function.iterate_zero],
+  rwa ←deriv_n
+end
+
+lemma deriv_n_add : (deriv_n (p+q) n) = (deriv_n p n) + (deriv_n q n) :=
+begin
+  induction n with n ih,
+  { simp only [zeroth_deriv], },
+  { rw [←deriv_n_succ, ih, polynomial.derivative_add, deriv_n_succ, deriv_n_succ] }
+end
+
+end semiring
+
+section comm_ring
+variable [comm_ring R]
+variables (f p q: polynomial R) (n k : ℕ)
+
+
+lemma deriv_n_coeff : ∀ m : ℕ, (deriv_n f k).coeff m = (∏ i in finset.range k, (m+k-i)) * (f.coeff (m+k)) :=
+begin
+  induction k with k ih,
+  { simp only [add_zero, forall_const, one_mul, range_zero, eq_self_iff_true,
+      zeroth_deriv, prod_empty] },
+
+  { intro m, rw ←deriv_n_succ,
+    rw [polynomial.coeff_derivative, ih (m+1), prod_range_succ],
+    simp only [nat.cast_succ, succ_eq_add_one],
+    conv_rhs { rw [mul_assoc, mul_comm] },
+    have triv : (∏ (i : ℕ) in range k, (m + 1 + k - i : R)) = ∏ (x : ℕ) in range k, (m + (k + 1) - x),
+    { apply congr_arg, ext, ring },
+    rw triv,
+    replace triv : (m + 1 : R) = (m + (k+1) - k:R),
+    { rw add_sub_assoc, simp only [add_sub_cancel'], }, rw ←triv,
+      replace triv : f.coeff (m + 1 + k) = f.coeff (m + (k + 1)),
+    { apply congr_arg, ring},
+    rw triv }
+end
+
+end comm_ring
+
+section integral_domain
+variables [integral_domain R]
+variables (f p q: polynomial R) (n k : ℕ)
+
+
+lemma deriv_too_much : (deriv_n f (f.nat_degree + 1)) = 0 :=
+begin
+  ext,
+  rw deriv_n_coeff,
+  simp only [cast_one, cast_add, coeff_zero],
+
+  rw mul_eq_zero, right,
+  apply polynomial.coeff_eq_zero_of_nat_degree_lt, linarith,
+end
+
+
+lemma deriv_n_poly_prod :
```

Comment:
```suggestion
lemma deriv_n_mul :
```

(also see my other comment about bikeshedding the name of the definition)

Reply:
this name was changed as per suggestion.